### PR TITLE
Create consistent performance iscsi disk support

### PR DESCRIPTION
**Pull request by jianqiu**
As SL no longer provided the ordering of legacy iscsi disk, softlayer-go needs to make corresponding changes to support consistent performance iscsi disk.
In order to enhance the efficiency of calling softlayer rest api, I created a new kind of http request in softlayer client. Several query methods in softlayer services had been added to leveraging filter function. For example, in production environment, there are a fair mount of storage, itemprice, blockdevicetemplates, etc. So it is better to filtering objects than traversing the result set multiple times.

New http request with objectFilter parameter in softlayer_client.go:

```go
func (slc *softLayerClient) DoRawHttpRequestWithObjectFilterAndObjectMask(path string, masks []string, filters string, requestType string, requestBody *bytes.Buffer) ([]byte, error) {
	url := fmt.Sprintf("https://%s:%s@%s/%s", slc.username, slc.apiKey, SOFTLAYER_API_URL, path)
	url += "?objectFilter=" + filters
	url += "&objectMask=filteredMask["
	for i := 0; i < len(masks); i++ {
		url += masks[i]
		if i != len(masks)-1 {
			url += ";"
		}
	}
	url += "]"
	return slc.makeHttpRequest(url, requestType, requestBody)
}
```

SoftLayer_Product_Package_Service.go
`+ GetItemPricesBySize(packageId int, size int) ([]datatypes.SoftLayer_Item_Price, error)`

SoftLayer_Account_Service.go
`+ GetIscsiNetworkStorageWithFilter(filter string) ([]datatypes.SoftLayer_Network_Storage, error)`

### CODE REVIEW

**Comment by maximilien** [comment]
@jianqiu,

In future, please use ``` to format any code in comments. You can also use single `quote` like this for simple one line or code inline.

For multiple line, use three ` like this:
```
multiple
line 
as 
code
```

File: services/softlayer_network_storage.go [hunk 1]
```diff
@@ -38,20 +38,33 @@ func (slns *softLayer_Network_Storage_Service) CreateIscsiVolume(size int, locat
 		return datatypes.SoftLayer_Network_Storage{}, errors.New("Cannot create negative sized volumes")
 	}
 
-	iscsiVolumeItemId, err := slns.getIscsiVolumeItemIdBasedOnSize(size)
-	if err != nil {
+	sizeItemPriceId, err := slns.getIscsiVolumeItemIdBasedOnSize(size)
+	iopsItemPriceId, err := slns.getPerformanceStorageItemPriceIdByIops(size)
+
+	/*if err != nil {
```
Comment (maximilien) [inline services/softlayer_network_storage.go]: Why is this line uncommented? I would expect a check for error after each call, so something like: 

```go
sizeItemPriceId, err := slns.getIscsiVolumeItemIdBasedOnSize(size)
if err != nil {
  return datatypes.SoftLayer_Network_Storage{}, err 
}

iopsItemPriceId, err := slns.getPerformanceStorageItemPriceIdByIops(size)
if err != nil {
  return datatypes.SoftLayer_Network_Storage{}, err 
}
```

Can you explain why you did not have check?

File: services/softlayer_network_storage.go [hunk 2]
```diff
@@ -38,20 +38,33 @@ func (slns *softLayer_Network_Storage_Service) CreateIscsiVolume(size int, locat
...
+			datatypes.SoftLayer_Item_Price{
+				Id: BLOCK_ITEM_PRICE_ID,
 			},
 		},
-		PackageId: NETWORK_STORAGE_PACKAGE_ID,
+		PackageId: NETWORK_PERFORMANCE_STORAGE_PACKAGE_ID,
+		Quantity:  1,
 	}
 
 	productOrderService, _ := slns.client.GetSoftLayer_Product_Order_Service()
```
Comment (maximilien) [inline services/softlayer_network_storage.go]: Why are we ignoring `err` here? We should never do so. If there is an `err` then we should fail.

File: services/softlayer_network_storage.go [hunk 3]
```diff
@@ -63,7 +76,7 @@ func (slns *softLayer_Network_Storage_Service) CreateIscsiVolume(size int, locat
 	var iscsiStorage datatypes.SoftLayer_Network_Storage
 
 	for i := 0; i < CREATE_ISCSI_VOLUME_MAX_RETRY_TIME; i++ {
-		iscsiStorage, err = slns.findIscsiVolumeIdByOrderId(receipt.OrderId)
+		iscsiStorage, err = slns.findIscsiVolumeId(receipt.OrderId)
 		if err == nil {
 			break
 		} else if i == CREATE_ISCSI_VOLUME_MAX_RETRY_TIME-1 {
```
Comment (maximilien) [inline services/softlayer_network_storage.go]: Same again here. We cannot just ignore errors, unless we can do something to recover.

File: services/softlayer_network_storage.go [hunk 4]
```diff
@@ -77,8 +90,9 @@ func (slns *softLayer_Network_Storage_Service) CreateIscsiVolume(size int, locat
 }
 
 func (slns *softLayer_Network_Storage_Service) DeleteIscsiVolume(volumeId int, immediateCancellationFlag bool) error {
+	ObjectFilter := string(`{"iscsiNetworkStorage":{"id":{"operation":` + strconv.Itoa(volumeId) + `}}}`)
 	accountService, _ := slns.client.GetSoftLayer_Account_Service()
-	iscsiStorages, _ := accountService.GetIscsiNetworkStorage()
+	iscsiStorages, _ := accountService.GetIscsiNetworkStorageWithFilter(ObjectFilter)
```
Comment (maximilien) [inline services/softlayer_network_storage.go]: Same deal here again. We cannot ignore errors.

File: services/softlayer_network_storage.go [hunk 5]
```diff
@@ -127,19 +141,17 @@ func (slns *softLayer_Network_Storage_Service) GetIscsiVolume(volumeId int) (dat
 
 // Private methods
 
-func (slns *softLayer_Network_Storage_Service) findIscsiVolumeIdByOrderId(orderId int) (datatypes.SoftLayer_Network_Storage, error) {
+func (slns *softLayer_Network_Storage_Service) findIscsiVolumeId(orderId int) (datatypes.SoftLayer_Network_Storage, error) {
+	ObjectFilter := string(`{"iscsiNetworkStorage":{"billingItem":{"orderItem":{"order":{"id":{"operation":` + strconv.Itoa(orderId) + `}}}}}}`)
 	accountService, _ := slns.client.GetSoftLayer_Account_Service()
```
Comment (maximilien) [inline services/softlayer_network_storage.go]: Same again

**Comment by maximilien** [comment]
All, too many issues. Please address and make sure all errors (I mean, all errors) are dealt with. This means never ever use `_` unless you have a very good reason---which you should document as a comment.

There are only two acceptable cases: 

1. you check for error and recover right then in there
2. you check for error and return it so it can be dealt with by the caller if the method or function.

I will be extra vigilant in will use @cppforlife vetting extension to vet previous code. Seems that I missed some of the old storage code. Let's correct that.

Thanks for you cooperation. Best,

@maximilien

**Comment by maximilien** [comment]
Since there are lots of changes and fixes to be addressed, I am closing this. 

Please address all comments and update and resubmit.

Thanks.

**Comment by jianqiu** [comment]
@maximilien 
Thanks for your explanation in a great detail, I had updated the codes, and submitted a new PR

**Pull request closed without merging**